vote/src/vote_account.rs

```diff
@@ -891,4 +992,153 @@ mod tests {
             }
         }
     }
+
+    #[test]
+    fn test_clone_and_filter_for_alpenglow_truncates() {
+        let mut rng = rand::thread_rng();
+        let current_limit = 3000;
+        let (vote_accounts, identity_balances) =
+            new_staked_vote_accounts(&mut rng, current_limit, current_limit, None);
+        // All vote accounts should be returned if the limit is high enough.
+        let filtered = vote_accounts.clone_and_filter_for_alpenglow(
+            current_limit + 500,
+            1,
+            &identity_balances,
+        );
+        assert_eq!(filtered.len(), vote_accounts.len());
+
+        // If the limit is smaller than number of accounts, truncate it.
+        let lower_limit = current_limit - 1000;
+        let filtered =
+            vote_accounts.clone_and_filter_for_alpenglow(lower_limit, 1, &identity_balances);
+        assert!(filtered.len() <= lower_limit);
+        // Check that the filtered accounts are the same as the original accounts.
+        for (pubkey, (_, vote_account)) in filtered.vote_accounts.iter() {
+            assert_eq!(vote_accounts.get(pubkey), Some(vote_account));
+        }
+        // Check that the stake in any filtered account is higher than truncated accounts.
+        let min_stake = filtered
+            .vote_accounts
+            .iter()
+            .map(|(_, (stake, _))| *stake)
+            .min()
+            .unwrap();
+        for (pubkey, (stake, _vote_account)) in vote_accounts.vote_accounts.iter() {
+            if *stake < min_stake {
+                assert!(filtered.get(pubkey).is_none());
+            }
+        }
+    }
+
+    #[test]
+    fn test_clone_and_filter_for_alpenglow_filters_non_alpenglow() {
+        let mut rng = rand::thread_rng();
+        let num_alpenglow_nodes = 2000;
+        // Check that non-alpenglow accounts are kicked out, 2000 accounts with bls pubkey, 1000 accounts without.
+        let num_nodes = num_alpenglow_nodes + 1000;
+        let (vote_accounts, identity_balances) =
+            new_staked_vote_accounts(&mut rng, num_nodes, num_alpenglow_nodes, None);
+        let new_limit = num_alpenglow_nodes + 500;
+        let filtered =
+            vote_accounts.clone_and_filter_for_alpenglow(new_limit, 1, &identity_balances);
+        assert!(filtered.len() <= num_alpenglow_nodes);
```

Comment:
should this be strictly `==` ?

Reply:
We could, because when the limit is 2000, we will drop everyone with the same stake as the 2001st one (given that they all have enough balance and BLS pubkey), so that can make the numbers sightly below 2000 when you assign stakes randomly.